### ISSUE TITLE
[BUG] Snowcat doesn't infer a type from equality

### ISSUE THREAD

**Issue by andrey-kuprianov** [informalsystems/apalache#864]
## Description

Snowcat complains about a polymorphic type when the type should be clear from equality:

```sh
Checker options: filename=snowcat_bug.tla, init=, next=, inv=     I@09:33:05.239
...
PASS #13: PostTypeCheckerSnowcat                                  I@09:33:06.414
 > Running Snowcat .::.                                           I@09:33:06.414
[unknown location]: Expected a concrete type of operator Empty$1, found polymorphic type: (() => Set(c)) E@09:33:06.421
 > Snowcat asks you to fix the types. Meow.                       I@09:33:06.421
Checker has found an error                                        I@09:33:06.422
It took me 0 days  0 hours  0 min  1 sec                          I@09:33:06.422
Total time: 1.185 sec                                             I@09:33:06.423
EXITCODE: ERROR (12)
```

## Input specification

Here is the MWE:

```tla
------------------------------------ MODULE snowcat_bug -----------------------------

\* @type: () => Int -> a;
EmptySeq == 
    LET (* @type: Set(Int); *) EmptyInt == {} IN 
    LET (* @type: Set(a); *) EmptyA == {} IN 
        [ i \in EmptyInt |-> CHOOSE j \in EmptyA: TRUE ]

VARIABLE 
\* @type: Int -> Int;
seq

Init == seq = EmptySeq

Next == UNCHANGED seq

===============================================================================
```

## The command line parameters used to run the tool

```sh
apalache check snowcat_bug.tla
```

## System information

- Apalache version : 0.15.7 build e6eda01
- OS: Ubuntu 20.04.2 LTS
- JDK version: openjdk version "15.0.2" 2021-01-19

**Comment by konnov** [comment]
I think it is related to #869. Working on a fix

**Comment by konnov** [comment]
This should be fixed by #953

**Comment by konnov** [comment]
Fixed via #953
